osu.Game/Screens/SelectV2/BeatmapCarousel.cs

```diff
@@ -471,16 +471,9 @@ protected override float GetPanelXOffset(Drawable panel)
 
         private ScheduledDelegate? loadingDebounce;
 
-        /// <summary>
-        /// We need to track this state locally since `FilterCriteria` is passed by reference and not accurate.
-        /// It should really be a struct.
-        /// </summary>
-        private bool splitOutDifficulties;
-
         public void Filter(FilterCriteria criteria)
         {
-            bool resetDisplay = criteria.SplitOutDifficulties != splitOutDifficulties;
-            splitOutDifficulties = criteria.SplitOutDifficulties;
+            bool resetDisplay = grouping.BeatmapSetsGroupedTogether != BeatmapCarouselFilterGrouping.ShouldGroupBeatmapsTogether(criteria);
```

Comment:
This makes me very uncomfortable because `grouping.BeatmapSetsGroupedTogether` is basically set on another thread as part of a fire-and-forget async process with indeterminate timing. Grouping runs as the last filter in queue, too. What is the guarantee that this is going to be in a valid state for reading when this is queried, rather than mid-operation when the value of it might not reflect reality?

I tried to insert an arbitrary sleep before the value of that property is set and couldn't in principle break it but I also am not sure that I *tried hard enough* to break it.

Reply:
I've kinda been avoiding the haphazardness of the `async` nature of the filter process because I'm not yet sure how to resolve it, but your concerns are valid.

This isn't the only usage which is haphazard. Accessing things like `SetItems` and `GroupItems` have the same fallacy. If possible, I'd like to turn a blind eye to this for the initial release. I do have this tracked.

![Things 2025-05-29 at 06 59 01](https://github.com/user-attachments/assets/f2dfa185-cee1-4b53-8fa2-cdb0f3a1c64f)

Reply:
I was going to suggest some alternative solution but then looked at usages of `BeatmapSetsGroupedTogether`, saw that it gets read in five other places that I can't remotely tell if they're safe reads, and decided to just give up for now.